xontrib/vox.py

```diff
@@ -1,77 +1,99 @@
+                             'by default)')
+
+    activate = subparsers.add_parser('activate', aliases=['workon', 'enter'], help='Activate virtual environment')
+    activate.add_argument('name', metavar='ENV', 
+                        help='The environment to activate')
+    subparsers.add_parser('deactivate', aliases=['exit'], help='Deactivate current virtual environment')
+    subparsers.add_parser('list', aliases=['ls'], help='List all available environments')
+    remove = subparsers.add_parser('remove', aliases=['rm', 'delete', 'del'], help='Remove virtual environment')
+    remove.add_argument('names', metavar='ENV', nargs='+',
+                        help='The environments to remove')
+    subparsers.add_parser('list', aliases=['help'], help='Show this help message')
+
+    aliases = {
+        'workon': 'activate',
+        'enter': 'activate',
+        'exit': 'deactivate',
+        'ls': 'list',
+        'rm': 'remove',
+        'delete': 'remove',
+        'del': 'remove',
+    }
 
     def __init__(self):
-        """Ensure that $VIRTUALENV_HOME is defined and declare the available vox commands"""
-
         self.vox = _voxapi.Vox()
 
-        self.commands = {
-            ('new',): self.create_env,
-            ('activate', 'workon', 'enter'): self.activate_env,
-            ('deactivate', 'exit'): self.deactivate_env,
-            ('list', 'ls'): self.list_envs,
-            ('remove', 'rm', 'delete', 'del'): self.remove_envs,
-            ('help', '-h', '--help'): self.show_help
-        }
-
     def __call__(self, args, stdin=None):
         """Call the right handler method for a given command."""
 
-        if not args:
-            self.show_help()
-            return None
-
-        command_name, params = args[0], args[1:]
-
-        try:
-            command = [
-                self.commands[aliases] for aliases in self.commands
-                if command_name in aliases
-            ][0]
-
-            command(*params)
-
-        except IndexError:
-            print('Command "%s" doesn\'t exist.\n' % command_name)
-            self.print_commands()
+        args = self.parser.parse_args(args)
+        cmd = self.aliases.get(args.command, args.command)
+        if cmd is None:
+            self.parser.print_usage()
+        else:
+            getattr(self, 'cmd_'+cmd)(args, stdin)
 
-    def create_env(self, name):
+    def cmd_create(self, args, stdin=None):
         """Create a virtual environment in $VIRTUALENV_HOME with python3's ``venv``.
-
-        Parameters
-        ----------
-        name : str
-            Virtual environment name
         """
         print('Creating environment...')
-        self.vox.create(name)
+        self.vox.create(args.name)
         msg = 'Environment {0!r} created. Activate it with "vox activate {0}".\n'
         print(msg.format(name))
 
-    def activate_env(self, name):
+    def cmd_activate(self, args, stdin=None):
```

Comment:
They're looked up dynamically in `__call__()`. I just use `cmd_` as a prefix to denote command-handling methods. (Like having `do_` in front of event handlers.)